src/session_state.rs

```diff
@@ -1,30 +1,34 @@
 /// This module represents the session state of an MQTT communication session.
 ///
 ///
+use crate::QoS;
 use embedded_nal::IpAddr;
-use heapless::{String, Vec};
+use heapless::{String, Vec, LinearMap};
 
 use embedded_time::{
     duration::{Extensions, Seconds},
-    Clock, Instant,
+    Clock as HwClock,
```

Comment:
Nit: Here and throughout, let's keep `embedded_time::Clock` defined as either `Clock` or fully specified as `embedded_time::Clock`. Using `Clock as HwClock` just makes the code harder to follow for someone not familiar with the codebase in my opinion.

If we're hitting naming issues, let's make the template parameter `Clock: embedded_time::Clock`